src/renderers/dom/shared/ReactDOMComponent.js

```diff
@@ -166,19 +166,17 @@ function assertValidProps(component, props) {
     );
     invariant(
       typeof props.dangerouslySetInnerHTML === 'object' &&
-      HTML in props.dangerouslySetInnerHTML &&
-      (!props.dangerouslySetInnerHTML[HTML] || typeof props.dangerouslySetInnerHTML[HTML] === 'string'),
-      '`props.dangerouslySetInnerHTML` must be in the form `{__html: ...}`. ' +
-      'Please visit https://fb.me/react-invariant-dangerously-set-inner-html ' +
-      'for more information.'
-    );
-    invariant(
-      typeof props.dangerouslySetInnerHTML === 'object' &&
       HTML in props.dangerouslySetInnerHTML,
       '`props.dangerouslySetInnerHTML` must be in the form `{__html: ...}`. ' +
       'Please visit https://fb.me/react-invariant-dangerously-set-inner-html ' +
       'for more information.'
     );
+    if(props.dangerouslySetInnerHTML[HTML]) {
```

Comment:
Embarrassingly, I did not succeed in combining this check with the one in the `warning` invocation below.

While the following passed the tests, I wasn't sure it's actually correct, plus it didn't read well:

```javascript
warning(
  !props.dangerouslySetInnerHTML[HTML] ||
  props.dangerouslySetInnerHTML[HTML].toString !== Object.prototype.toString,
  '`props.dangerouslySetInnerHTML` should be a string.'
);
```

Boolean logic can be surprisingly tricky.

¹ cf. comment above: I'm currently very tired - but figured I better get this done right away rather than risk procrastination.